src/main/java/com/zaxxer/hikari/pool/ProxyConnection.java

```diff
@@ -170,7 +171,7 @@ final void untrackStatement(final Statement statement)
 
    final void markCommitStateDirty()
    {
-      if (isAutoCommit) {
+      if (isAutoCommit || isReadOnly) {
```

Comment:
wrt: expensive, I had created issue https://github.com/brettwooldridge/HikariCP/issues/576 for same reason.
IMO with narrower validation window and many stability improvements later, it is 'ok' to live with that 'rare' problem as you mentioned there till better solution.